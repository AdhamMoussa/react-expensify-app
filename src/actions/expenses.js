import uuid from "uuid";

export const addExpense = ({
  description = "",
  amount = 0,
  note = "",
  createdAt = 0
} = {}) => {
  return {
    type: "ADD_EXPENSE",
    newExpense: {
      id: uuid(),
      description,
      amount,
      note,
      createdAt
    }
  };
};

export const editExpense = (id, updatedData) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updatedData
  };
};

export const removeExpense = (id) => {
  return {
    type: 'REMOVE_EXPENSE',
    id
  };
};
