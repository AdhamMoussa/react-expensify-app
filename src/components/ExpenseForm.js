import React from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

class ExpenseForm extends React.Component {
  state = {
    description: this.props.editedExpense
      ? this.props.editedExpense.description
      : "",
    amount: this.props.editedExpense ? this.props.editedExpense.amount : "",
    note: this.props.editedExpense ? this.props.editedExpense.note : "",
    createdAt: this.props.editedExpense
      ? this.props.editedExpense.createdAt
      : moment(),
    datePickerFocused: false,
    descriptionError: null,
    amountError: null
  };
  handleDescriptionChange = e => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };
  handleAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/))
      this.setState(() => ({ amount }));
  };
  handleNoteChange = e => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };
  handleFormSubmit = e => {
    e.preventDefault();
    // handle description error
    if (!this.state.description) {
      this.setState(() => ({
        descriptionError: "Please insert a description of your expense"
      }));
    } else if (this.state.description && this.state.descriptionError) {
      this.setState(() => ({ descriptionError: null }));
    }
    // handle amount error
    if (!this.state.amount) {
      this.setState(() => ({
        amountError: "Please insert an amount of your expense"
      }));
    } else if (this.state.amount && this.state.amountError) {
      this.setState(() => ({ amountError: null }));
    }
    // Submit
    if (this.state.description && this.state.amount) {
      this.setState(() => ({
        descriptionError: null,
        amountError: null
      }));
      this.props.onSubmitHandler({
        description: this.state.description,
        amount: parseFloat(this.state.amount),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };
  render() {
    return (
      <div>
        <form onSubmit={this.handleFormSubmit}>
          <input
            id="description-input"
            type="text"
            name="description"
            value={this.state.description}
            onChange={this.handleDescriptionChange}
            placeholder="Expense Description"
            autoFocus
          />
          <br />
          {this.state.descriptionError && <p>{this.state.descriptionError}</p>}
          <input
            id="amount-input"
            type="text"
            name="amount"
            value={this.state.amount}
            onChange={this.handleAmountChange}
            placeholder="Amount"
          />
          <br />
          {this.state.amountError && <p>{this.state.amountError}</p>}
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={date => {
              if (date) {
                this.setState({ createdAt: date });
              }
            }}
            focused={this.state.datePickerFocused}
            onFocusChange={({ focused }) =>
              this.setState({ datePickerFocused: focused })
            }
            numberOfMonths={1}
            isOutsideRange={day => false}
          />
          <br />
          <textarea
            id="note-input"
            name="note"
            value={this.state.note}
            onChange={this.handleNoteChange}
            placeholder="Expense Note"
          />
          <br />
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
