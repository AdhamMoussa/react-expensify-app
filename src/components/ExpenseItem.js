import React from "react";
import { Link } from 'react-router-dom';

const ExpenseItem = (props) => {
  return (
    <div>
      <Link to={`/edit/${props.id}`}>
        <h3>Expense: {props.description}</h3>
      </Link>
      <p>Amount: {props.amount}</p>
      <p>Date: {props.createdAt.format('MMM Do, YYYY')}</p>
    </div>
  );
};

export default ExpenseItem;
