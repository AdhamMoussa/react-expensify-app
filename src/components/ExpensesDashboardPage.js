import React from 'react';
import ExpensesList from './ExpensesList';

const ExpensesDashboardPage = (props) => {
  return (
    <div>
      <h2>Your Expenses</h2>
      <ExpensesList />
    </div>
  )
};


export default ExpensesDashboardPage;
