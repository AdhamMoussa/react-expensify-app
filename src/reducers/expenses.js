const defaultExpensesState = [];

const expensesReducer = (state = defaultExpensesState, action) => {
  switch (action.type) {
    case "ADD_EXPENSE":
      return [...state, action.newExpense];
    case "EDIT_EXPENSE":
      return state.map(
        expense =>
          expense.id === action.id
            ? { ...expense, ...action.updatedData }
            : expense
      );
    case "REMOVE_EXPENSE":
      return state.filter(expense => expense.id !== action.id);
    default:
      return state;
  }
};

export default expensesReducer;
