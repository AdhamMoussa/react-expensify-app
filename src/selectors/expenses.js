import moment from 'moment';

const getExpensesView = (expenses, filters) => {
  return expenses.filter((expense) => {
    const textMatch = expense.description.toLowerCase().includes(filters.text.toLowerCase());
    const startDateMatch = filters.startDate ? expense.createdAt.isSameOrAfter(filters.startDate) : true;
    const endDateMatch = filters.endDate ? expense.createdAt.isSameOrBefore(filters.endDate) : true;

    return textMatch && startDateMatch && endDateMatch;
  }).sort((a, b) => {
    if (filters.sortBy === 'date') {
      return a.createdAt.isAfter(b.createdAt) ? -1 : 1;
    } else if (filters.sortBy === 'amount') {
      return a.amount > b.amount ? -1 : 1;
    }
  });
};

export default getExpensesView;
