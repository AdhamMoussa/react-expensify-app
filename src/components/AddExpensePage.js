import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { addExpense } from '../actions/expenses';


export const AddExpensePage = (props) => {
  const handleFormSubmit = (expense) => {
    props.onSubmit(expense);
    props.history.push('/');
  };
  return (
    <div>
      <h2>Add Expense</h2>
      <ExpenseForm onSubmitHandler={handleFormSubmit} />
    </div>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    onSubmit: (expense) => dispatch(addExpense(expense))
  };
};

export default connect(undefined, mapDispatchToProps)(AddExpensePage);
