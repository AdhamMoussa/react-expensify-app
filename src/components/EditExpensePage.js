import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

export const EditExpensePage = props => {
  const onSubmitHandler = expense => {
    props.onSubmit(props.expense.id, expense);
    props.history.push("/");
  };
  const onRemoveHandler = () => {
    props.onRemoveClick(props.expense.id);
    props.history.push("/");
  };
  return (
    <div>
      <h2>Edit Expense</h2>
      <ExpenseForm
        onSubmitHandler={onSubmitHandler}
        editedExpense={props.expense}
      />
      <button id="remove-btn" onClick={onRemoveHandler}>
        Remove
      </button>
    </div>
  );
};

const mapDispatchToProps = dispatch => {
  return {
    onSubmit: (id, expense) => dispatch(editExpense(id, expense)),
    onRemoveClick: id => dispatch(removeExpense(id))
  };
};

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(exp => exp.id === props.match.params.id)
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(EditExpensePage);
