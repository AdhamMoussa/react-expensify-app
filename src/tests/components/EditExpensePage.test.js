import React from "react";
import { shallow } from "enzyme";
import { EditExpensePage } from "../../components/EditExpensePage";
import ExpenseForm from "../../components/ExpenseForm";
import expenses from "../fixtures/demoExpenses";

test("should render EditExpensePage correctly", () => {
  const wrapper = shallow(<EditExpensePage />);
  expect(wrapper).toMatchSnapshot();
});

test("should call onSubmitHandler on form submit", () => {
  const onSubmitSpy = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <EditExpensePage
      onSubmit={onSubmitSpy}
      history={history}
      expense={expenses[0]}
    />
  );
  wrapper.find(ExpenseForm).prop("onSubmitHandler")(expenses[0]);
  expect(onSubmitSpy).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
});

test("should call onRemoveClick on remove btn click", () => {
  const onRemoveSpy = jest.fn();
  const history = { push: jest.fn() };
  const wrapper = shallow(
    <EditExpensePage
      onRemoveClick={onRemoveSpy}
      expense={expenses[1]}
      history={history}
    />
  );
  wrapper.find("#remove-btn").simulate("click");
  expect(onRemoveSpy).toHaveBeenLastCalledWith(expenses[1].id);
  expect(history.push).toHaveBeenLastCalledWith("/");
});
