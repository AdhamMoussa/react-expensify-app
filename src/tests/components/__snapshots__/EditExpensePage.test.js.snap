// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditExpensePage correctly 1`] = `
<div>
  <h2>
    Edit Expense
  </h2>
  <ExpenseForm
    onSubmitHandler={[Function]}
  />
  <button
    id="remove-btn"
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
